src/methods/join.js

```diff
@@ -807,18 +831,70 @@ export class JoinImplementation {
     }
   }
 
+  // Joins property bindings at a join point. Returns joined bindings and a set of leaked
+  // objects, based on the binding for the property "_isLeaked", which marks objects as
+  // leaked.
   joinPropertyBindings(
     realm: Realm,
     joinCondition: AbstractValue,
     m1: PropertyBindings,
     m2: PropertyBindings,
     c1: CreatedObjects,
-    c2: CreatedObjects
-  ): PropertyBindings {
-    let join = (b: PropertyBinding, d1: void | Descriptor, d2: void | Descriptor) => {
+    c2: CreatedObjects,
+    g1: Generator,
+    g2: Generator
+  ): [PropertyBindings, IntermediateObjects] {
+    // Joins two property bindings of type "_isLeaked". If an object is leaked
+    // on one side, it is also leaked on the other side. Returns [x, y] where
+    // x is the joined descriptor, and y is set to the IntermediateObjectValue consisting of generator and object if the object leaked.
+    let joinLeakedObjectPropertyAndAdvise = (
+      b: PropertyBinding,
+      d1: void | Descriptor,
+      d2: void | Descriptor
+    ): [Descriptor | void, IntermediateObjectValue | void] => {
+      if (d1 === undefined) {
+        if (c2.has(b.object)) return [d2, undefined]; // no join, no leak advise
+
+        invariant(m2.has(b)); // The alternative path must be the source of leaking
+        return [d2, { object: b.object, generator: g1 }];
+      } else if (d2 === undefined) {
+        if (c1.has(b.object)) return [d1, undefined]; // no join, no leak
+
+        invariant(m1.has(b)); // The consequent path must be the source of leaking
+        return [d1, { object: b.object, generator: g2 }];
+      } else {
+        // Make the leak unconditional
+        let isTrue = bv => bv instanceof BooleanValue && bv.value;
+        let _d1Value = isTrue(d1.value);
```

Comment:
We already have !d1.value.mightNotBeTrue(). Also, the _ prefix is used for private object properties and using it for a local name seems strange. d1ValueIsTrue would be a better name.